### ISSUE TITLE
Add hashsum command, with --download, to rc API

### ISSUE THREAD

**Issue by Jwink3101** [rclone/rclone#7569]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/rc-endpoint-for-hashsum-download/43965 and a bit indirect, but https://forum.rclone.org/t/core-command-does-not-respect-flags-from-rcd/43970/2

#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.65.1
- os/version: darwin 14.2.1 (64 bit)
- os/kernel: 23.2.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.21.5
- go/linking: dynamic
- go/tags: cmount
```

#### What problem are you are trying to solve?

There is no way to do `hashsum --download` in the RC api. `operations/list` and `operations/stat` can do hashes if and only if they are native to the remote. Features such as `--download`, `--base64`, and `--checkfile` (and probably more) are not supported.

To be fair, this may fall into the realm of something you can do with rclone but do not need to do it all. You could stream the and compute the hashes and/or compare with a different programming tool but I don't think that is a fair metric.

#### How do you think rclone should be changed to solve that?

Add this command to the API. Probably something like `operations/hashsum`

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Please give this a go

[v1.66.0-beta.7622.62a758961.fix-7569-rc-hashsum](https://beta.rclone.org/branch/fix-7569-rc-hashsum/v1.66.0-beta.7622.62a758961.fix-7569-rc-hashsum/) on branch [fix-7569-rc-hashsum](https://github.com/rclone/rclone/tree/fix-7569-rc-hashsum) (uploaded in 15-30 mins)

### operations/hashsum: Produces a hashsum file for all the objects in the path. {#operations-hashsum}

Produces a hash file for all the objects in the path using the hash
named.  The output is in the same format as the standard
md5sum/sha1sum tool.

This takes the following parameters:

- fs - a remote name string e.g. "drive:" for the source, "/" for local filesystem
- hashType - type of hash to be used
- download - check by downloading rather than with hash (boolean)
- base64 - output the hashes in base64 rather than hex (boolean)

If you supply the download flag, it will download the data from the
remote and create the hash on the fly. This can be useful for remotes
that don't support the given hash or if you really want to check all
the data.

Note that if you wish to supply a checkfile to check hashes against
the current files then you should use operations/check instead of
operations/hashsum.

Returns:

- hashsum - array of strings of the hashes
- hashType - type of hash used

Example:

    $ rclone rc --loopback operations/hashsum fs=bin hashType=MD5 download=true base64=true
    {
        "hashType": "md5",
        "hashsum": [
            "WTSVLpuiXyJO_kGzJerRLg==  backend-versions.sh",
            "v1b_OlWCJO9LtNq3EIKkNQ==  bisect-go-rclone.sh",
            "VHbmHzHh4taXzgag8BAIKQ==  bisect-rclone.sh",
        ]
    }

See the [hashsum](/commands/rclone_hashsum/) command for more information on the above.

**Authentication is required for this call.**

**Comment by Jwink3101** [comment]
This does not allow for requesting a single file. For example, on the CLI, I can do:

    $ rclone hashsum quickxor onedrive:Pictures/Photos/photo.jpg

but on the rc side:

```
rclone rc --loopback operations/hashsum fs='onedrive:Pictures/Photos/photo.jpg' hashType=quickxor
{
        "error": "loopback: call failed: is a file not a directory",
        "path": "operations/hashsum",
        "status": 500
}
```
Following other idioms from the rc API, I tried adding `remote`

```
rclone rc --loopback operations/hashsum fs='onedrive:Pictures/Photos/' remote='photo.jpg' hashType=quickxor
```

and it just hangs (not surprising)

**Comment by ncw** [comment]
What the command line does is put in a filter when you specify a file. So you could do that with _filter.

I'll look at making remote= work though that should be easy enough.

There are probably other rc calls like that.

**Comment by Jwink3101** [comment]
> There are probably other rc calls like that.

Yes. Many of the operations are like that. Probably the closest would be `operations/stat`

**Comment by ncw** [comment]
Give this a go :-)

[v1.66.0-beta.7628.d787b38e7.fix-7569-rc-hashsum](https://beta.rclone.org/branch/fix-7569-rc-hashsum/v1.66.0-beta.7628.d787b38e7.fix-7569-rc-hashsum/) on branch [fix-7569-rc-hashsum](https://github.com/rclone/rclone/tree/fix-7569-rc-hashsum) (uploaded in 15-30 mins)

**Comment by Jwink3101** [comment]
That worked! Was the idea not to create a `remote` key? I guess if working on a file is not the main operation, that wouldn't be the right idiom but it's a tossup.

I also got it to work manually with filters as you noted before

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.66
